doc/neps/nep-0041-improved-dtype-support.rst

```diff
@@ -0,0 +1,706 @@
+
+The large need to solve these issues has driven the scientific community
+to create work-arounds in multiple projects implementing physical units as an
+array-like class instead of a datatype, which would generalize better across
+multiple array-likes (Dask, pandas, etc.).
+Already, Pandas has made a push into the same direction with its
+extension arrays [pandas_extension_arrays]_ and undoubtedly
+the community would be best served if such new features could be common
+between NumPy, Pandas, and other projects.
+
+Scope
+^^^^^
+
+This NEP proposes to move ahead with the necessary creation of new dtype
+subclasses (see Phase I Implementation section),
+and start working on implementing current functionality within the new context
+Within the context of this NEP all development will be fully private API or
+use preliminary underscored names which must be changed in the future.
+Most of the internal and public API choices are part of a second Phase
+and will be discussed in more detail in the following NEPs 42 and 43.
+The initial implementation of this NEP will have little or no effect on users,
+but provides the necessary ground work for incrementally addressing the
+full rework.
+
+The implementation of this NEP and leading up to a large rework of how
+datatypes are defined in NumPy is expected to create small incompatibilities
+(see backward compatibility section).
+However, a transition requiring large code adaption is not anticipated and not
+within scope.
+
+Specifically, this NEP makes the following design choices which are discussed
+in more details in the detailed description section:
+
+1. Each datatype will be a class instance of a DType class, with most of the
+   datatype-specific logic being implemented
+   as special methods on the class. In the C-API, these correspond to specific
+   slots. In short ``type(np.dtype("f8"))`` will be a subclass of ``np.dtype``.
+   All the methods which are currently stored on the instance, should instead
+   be defined on the class as is normal in Python.
+   Storage information such as itemsize and byteorder can differ between
+   different dtype instances (e.g. "S3" vs. "S8") and will remain part of the instance.
+   This means that in the long run the current lowlevel access to dtype methods
+   will be removed (see ``ArrFunctions`` in NEP 40).
+
+2. The current NumPy scalars will *not* change, they will not be instances of
+   datatypes. This will also be true for new datatypes, scalars will not be
+   instances of a dtype (although ``isinstance(scalar, dtype)`` may be defined).
+
+With detailed technical decisions being followed up in NEP 42.
+
+Further, the public API will be designed in a way that is extensible in the future:
+
+3. All new API functions provided to the user will hide implementation details
+   as much as possible. The public API should be an identical, but limited,
+   version of the API used for the internal NumPy datatypes.
```

Comment:
Can we be clear about C vs Python API here?

Reply:
Right added the `C-`, Python is not very relevant for now.